### PR TITLE
feat(profiling): add profile summary

### PR DESCRIPTION
**Pull request by JonasBa**
Adds a top 10 longest application and system calls table (by self time). We currently have no way to link to this view, but we can figure that out in follow up PR. Also adds a ProfileGroupProvider that wraps the relevant routes so we can share this data across routes.

CleanShot 2022-05-05 at 10 29 58

### CODE REVIEW

File: static/app/utils/profiling/profile/utils.tsx [hunk 1]
```diff
@@ -113,3 +117,71 @@ export function wrapWithSpan<T>(parentSpan: Span | undefined, fn: () => T, optio
...
+
+type AnalyzeProfileResults = {
+  slowestApplicationCalls: CallTreeNode[];
+  slowestSystemCalls: CallTreeNode[];
+};
+
+export function getSlowestProfileCallsFromProfile(
+  profile: Profile
+): AnalyzeProfileResults {
+  const ApplicationCalls: CallTreeNode[] = [];
```
Comment (Zylphrex) [inline static/app/utils/profiling/profile/utils.tsx:137]: nit:
```suggestion
  const applicationCalls: CallTreeNode[] = [];
```

File: static/app/views/profiling/flamegraphSummary.tsx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            isLoading={state.type === 'loading'}
+            error={state.type === 'errored'}
+            data={functions.slowestApplicationCalls}
+            columnOrder={COLUMN_ORDER.map(key => COLUMNS[key])}
+            columnSortBy={[]}
+            grid={{renderBodyCell: renderFunctionCell}}
+            location={location}
+          />
+          <GridEditable
+            title="Slowest System Calls"
```
Comment (Zylphrex) [inline static/app/views/profiling/flamegraphSummary.tsx:95]:
```suggestion
            title={t('Slowest System Calls')}
```

File: static/app/views/profiling/profileGroupProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  data: T;
+  type: 'resolved';
+};
+
+type ErroredState = {
+  error: string;
+  type: 'errored';
+};
+
+type RequestState<T> = InitialState | LoadingState | ResolvedState<T> | ErroredState;
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/views/profiling/profileGroupProvider.tsx:25]: yeah, I think this should be some utility useAsync thing, seems weird that we are implementing it tbh

File: static/app/views/profiling/flamegraphSummary.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  column,
+  dataRow,
+}: ProfilingFunctionsTableCellProps) {
+  const value = dataRow[column.key];
+
+  switch (column.key) {
+    case 'self weight':
+      return <NumberContainer>{formatter(value)}</NumberContainer>;
+    case 'total weight':
+      return <NumberContainer>{formatter(value)}</NumberContainer>;
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/views/profiling/flamegraphSummary.tsx:144]: yeah, we should do that. I'll follow up in a different PR, I suspect this is going to impact a lot of parts where we format the code + the API of getDuration seems to be slightly different than what we use in the flamegraphs

File: static/app/views/profiling/flamegraph.tsx [hunk 5]
```diff
@@ -76,44 +43,19 @@ interface FlamegraphViewProps {
...
-      });
-
-    return () => {
-      api.clear();
-    };
-  }, [props.params.eventId, props.params.projectId, api, organization]);
-
   const onImport: ProfileDragDropImportProps['onImport'] = profiles => {
-    setRequestState({type: 'resolved', data: profiles});
+    setProfileGroup({type: 'resolved', data: profiles});
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/views/profiling/flamegraph.tsx:58]: Yeah, I think we probably want to have a way to set error too. I'd just keep as is for now, we need a useAsync hook :D

File: static/app/views/profiling/flamegraphSummary.tsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            data={functions.slowestApplicationCalls}
+            columnOrder={COLUMN_ORDER.map(key => COLUMNS[key])}
+            columnSortBy={[]}
+            grid={{renderBodyCell: renderFunctionCell}}
+            location={location}
+          />
+          <GridEditable
+            title="Slowest System Calls"
+            isLoading={state.type === 'loading'}
+            error={state.type === 'errored'}
```
Comment (Zylphrex) [inline static/app/views/profiling/flamegraphSummary.tsx:97]: The error here should be the error message you want to display

**Review by Zylphrex — APPROVED** [review]

**Pull request merged**